tests/run-tests

```diff
@@ -0,0 +1,166 @@
+#    idempotent.
+#
+#    destroy: tears down the VM. If this action isn't invoked, the VM
+#    is left running when the script exits and subsequent runs will
+#    reuse it.
+
+# Examples:
+#
+# A single run of the playbook typical during development. Leaves the
+# VM running after the script exits.
+#
+#    ./run-tests incremental
+#
+# The full test suite for use in CI.
+#
+#    ./run-tests ansible_test full idempotency destroy
+
+builddir="${HOME}/dlconfig"
+
+fail() {
+    echo $*
+    exit 1
+}
+
+cleanup() {
+    [[ -n "${builddir}" ]] && rm -rf ${builddir}  ${HOME}/secrets.yaml
+    echo "cleanup done"
+}
+
+trap cleanup EXIT
+
+destroy() {
+    vagrant destroy --force
+}
+
+ansible_test() {
+    # In a subshell so that the cd doesn't affect later actions
+    (
+        cd ${builddir}/ansible_collections/iridl/iridl &&
+        ansible-test sanity --docker
+    )
+}
+
+full() {
+    _playbook --build
+}
+
+incremental() {
+    _playbook
+}
+
+idempotency() {
+    echo starting idempotency test $(date)
+    output=$(_playbook)
+    playbook_status=$?
+    echo "$output" | grep '^dlserver1 .*changed=0 ' > /dev/null
+    idempotency_status=$?
+    if (( $playbook_status || $idempotency_status )); then
+        echo "$output"
+        echo idempotency test failed $(date)
+        return 1
+    fi
+    echo idempotency test succeeded $(date)
+}
+
+_install() {
+    if [[ ${guest} -eq 0 ]]; then
+      vagrant rsync
+      return
+    fi
+
+    if [[ ! -d "${builddir}" ]]; then
+        mkdir "${builddir}" || return 1
+        source /opt/datalib_venv3.12/bin/activate
+        ansible-galaxy collection install -p ${builddir} .. || return 1
+        deactivate
+        cp ${builddir}/ansible_collections/iridl/iridl/example/* ${builddir}
+        mv ${builddir}/secrets.yaml ${builddir}/..
+    fi
+}
+
+_up() {
+    # Try repeatedly to connect to the VM. If the ssh port isn't open
+    # when ansible attempts to connect (e.g. after the provisioner
+    # reboots the VM), then attempts to connect to it fail immediately
+    # rather than waiting for the connection timeout. The
+    # vagrant-reload plugin would be a cleaner way to handle this, but
+    # the version of vagrant I'm getting from ubuntu 18.04 is too old
+    # to support that.
+    vagrant up || return 1
+    echo waiting for VM
+    i=10
+    while (( i > 0 )); do
+        if vagrant ssh -- true; then
+            break
+        fi
+        sleep 10
+    done
+    if (( i > 0 )); then
+        echo VM is up
+    else
+        echo VM failed to come up
+        return 1
+    fi
+    # Update shared folder. Docs say this happens automatically, but
+    # it doesn't. Maybe my version of vagrant is too old.
+    vagrant rsync
+}
+
+_playbook() {
+    (
+        cd ${builddir} &&
+        ./run-ansible $*
+    )
+}
+
+cd $(dirname $0) || fail
+
+guest=0
+myname=$(id -un)
+if [[ "${myname}" == "vagrant" ]]; then
+  guest=1
+else
+  _up
```

Comment:
The old script called `_up` at the start of each of the other commands. Why did you move it here?

`./run-tests destroy full` is how I typically invoke the script. In this version, `_up` is called before `destroy`, so by the time we get to `full` the VM has already been provisioned and torn down, and there's no more VM to run the tests on.